eddymotion/model.py

```diff
@@ -127,21 +152,86 @@ def __init__(self, gtab, S0=None, mask=None, **kwargs):
                 "jac",
             )
         }
-        self._model = DipyTensorModel(gtab, **kwargs)
 
-    def fit(self, data, **kwargs):
+        # Create a TensorModel for each chunk
+        self._model_chunks = [
+            DipyTensorModel(gtab, **kwargs)
+            for _ in range(self._n_threads)
+        ]
+
+    @staticmethod
+    def fit_chunk(model_chunk, data_chunk):
         """Call model's fit."""
-        self._model = self._model.fit(data[self._mask, ...])
+        return model_chunk.fit(data_chunk)
 
-    def predict(self, gradient, step=None, **kwargs):
-        """Propagate model parameters and call predict."""
-        predicted = np.squeeze(
-            self._model.predict(
+    def fit(self, data, **kwargs):
+        """Fit the model chunk-by-chunk asynchronously."""
+        if self._mask is not None:
+            data = data[self._mask, ...]
+
+        # Split data into chunks of group of slices (axis=2)
+        data_chunks = np.split(data, self._n_threads, axis=2)
+
+        # Create a limited thread pool.
+        executor = concurrent.futures.ThreadPoolExecutor(
+            max_workers=self._n_threads,
+        )
+
+        loop = asyncio.get_event_loop()
+        fit_tasks = [
+            loop.run_in_executor(
+                executor,
+                self.fit_chunk,
+                self._model_chunks[i],
+                data_chunks[i]
+            )
+            for i in range(self._n_threads)
+        ]
+
+        try:
+            event_loop = asyncio.get_event_loop()
+            self._model_chunks = event_loop.run_until_complete(asyncio.gather(*fit_tasks))
```

Comment:
Indeed, it would be good to narrow down the kinds of errors we might expect here - but, given that there's no `except`, I assume this is to ensure that the `finally` block is executed in all cases.

The exception is then reraised unchanged. Is that it, Seb?

Reply:
Good catch! the try/finally is indeed useless here. I got inspired by the example in https://pymotw.com/3/asyncio/executors.html#threads that uses this to ensure that the event loop is closed at the very end. But when called in a jupyter notebook, it was raising ``Error: cannot close a running loop``. More specifically the problem is that `get_event_loop` gets an existing event loop that the call does not have permission to close.

I made a few changes in my code on my side where I am now creating a new event loop. For instance, the code for the fit looks like:
```python
with ThreadPoolExecutor(max_workers=self._n_threads) as executor:
    loop = asyncio.new_event_loop()

    fit_tasks = [
        loop.run_in_executor(
            executor,
            self.fit_chunk,
            self._model_chunks[i],
            data_chunks[i]
        )
        for i in range(self._n_threads)
    ]

    try:
        self._model_chunks = loop.run_until_complete(asyncio.gather(*fit_tasks))
    finally:
        loop.close()
```
In this way we have full control on the event loop and I use the `finally` clause to ensure that the created event loop is closed at the very end.

Alternatively, I could also keep the `get_event_loop` and remove the try/finally clauses and never call the `loop.close()`. WDYT?

I haven't pushed changes yet to github as @oesteban asked me to hold off fixing merge conflicts (https://github.com/nipreps/EddyMotionCorrection/pull/34#issuecomment-819542777)  but happy to push changes corresponding to one of the 2 alternatives when I get green light.